apps/omg_performance/lib/omg_performance/byzantine_events.ex

```diff
@@ -0,0 +1,146 @@
+# Copyright 2019 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.Performance.ByzantineEvents do
+  @moduledoc """
+  OMG network child chain server byzantine event test entrypoint. Setup and runs performance byzantine tests.
+
+  # Usage
+
+  See functions in this module for options available
+
+  ## start_dos_get_exits runs a test to get exit data for given 10 positions for 3 users
+
+  ```
+  mix run --no-start -e \
+    '
+      OMG.Performance.ByzantineEvents.Generators.stream_utxo_positions() |>
+      Enum.take(10) |> OMG.Performance.ByzantineEvents.start_dos_get_exits(3)
+    '
+  ```
+
+  __ASSUMPTIONS:__
+  This test should be run on testnet filled with transactions make sure you followed instructions in `docs/demo_05.md`
+  and `geth`, `omg_child_chain` and `omg_watcher` are running; and watcher is fully synced.
+
+  Expected result of running the above command should looks like:
+
+  ```
+  [
+    %{span_ms: 232000, corrects_count: 10, errors_count: 0},
+    %{span_ms: 221500, corrects_count: 10, errors_count: 0},
+    %{span_ms: 219900, corrects_count: 10, errors_count: 0},
+  ]
+  ```
+
+  where the sum of `corrects_count + errors_count` should equal to `length(positions)`.
+  If all passed position was unspent there should be no errors.
+  """
+
+  use OMG.Utils.LoggerExt
+
+  alias OMG.Eth
+  alias OMG.Performance.ByzantineEvents.DoSExitWorker
+  alias OMG.Performance.HttpRPC.WatcherClient
+  alias OMG.Utils.HttpRPC.Encoding
+
+  @type stats_t :: %{
+          span_ms: non_neg_integer(),
+          corrects_count: non_neg_integer(),
+          errors_count: non_neg_integer()
+        }
+
+  @watcher_url Application.get_env(:omg_performance, :watcher_url)
+  @micros_in_millisecond 1000
+
+  @doc """
+  For given utxo positions and given number of users start fetching exit data from Watcher.
+  User tasks are run asynchronously, each user receives the same positions list, shuffle its order
+  and call Watcher for exit data sequentially one position at a time.
+  """
+  @spec start_dos_get_exits([non_neg_integer()], [OMG.TestHelper.entity()], watcher_url: binary()) ::
+          stats_t()
+  def start_dos_get_exits(positions, dos_users, url \\ @watcher_url) do
+    1..dos_users
+    |> Enum.map(fn _ -> DoSExitWorker.get_exits_fun(positions, url) |> Task.async() end)
```

Comment:
I think we should break this line up? Looks like the [pipe is unwarranted here](https://github.com/lexmag/elixir-style-guide#anonymous-pipeline). Also maybe we can use the other form here if you don't want to break this out into a named function?

```suggestion
    |> Enum.map(fn _ -> Task.async(DosExitWorker, :get_exits_fun, [positions, url]) end)
```

Reply:
This is nice! Thank you ❗️

Reply:
Looks easy but I was unable to make it work... refactored a little for readability.
Hope it's better now.